RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc

```diff
@@ -68,25 +70,29 @@ void HGCalImagingAlgo::populate(const HGCRecHitCollection& hits){
 // with different input (reset should be called between events)
 void HGCalImagingAlgo::makeClusters()
 {
-
-  // used for speedy search
-  std::vector<KDTree> hit_kdtree(2*(maxlayer+1));
-
+  std::vector<std::vector<std::vector< KDNode> > >  layerClustersPerLayer(2*(maxlayer+2));
+  edm::SerialTaskQueue layersQueue;
   //assign all hits in each layer to a cluster core or halo
-  for (unsigned int i = 0; i <= 2*maxlayer+1; ++i) {
-    KDTreeBox bounds(minpos[i][0],maxpos[i][0],
-		     minpos[i][1],maxpos[i][1]);
-
-    hit_kdtree[i].build(points[i],bounds);
-
-    unsigned int actualLayer = i > maxlayer ? (i-(maxlayer+1)) : i; // maps back from index used for KD trees to actual layer
-
-    double maxdensity = calculateLocalDensity(points[i],hit_kdtree[i], actualLayer); // also stores rho (energy density) for each point (node)
-    // calculate distance to nearest point with higher density storing distance (delta) and point's index
-    calculateDistanceToHigher(points[i]);
-    findAndAssignClusters(points[i],hit_kdtree[i],maxdensity,bounds,actualLayer);
-  }
-  //make the cluster vector
+  tbb::parallel_for(size_t(0), size_t(2*maxlayer+2), [&](size_t i) {
```

Comment:
This needs to be done in the context of a `tbb::this_task_arena::isolate` call. See
https://software.intel.com/en-us/node/684814

This avoid accidently having another module's task run on-top of this module in the case where the `tbb::parallel_for` decides to task steal while waiting for other threads to complete the assigned work for the `for` loop.